### PR TITLE
NZSL-74 csv import from share

### PR DESCRIPTION
**Pull request by nora-ns**
Working csv import from share. 
It uses the existing csv import as a base, but then uses `csv.DictReader` rather than `csv.reader`.
Creation of GlossVideos is handed to a separate thread to run in the background, otherwise it would quickly cause request time outs.

### CODE REVIEW

File: signbank/dictionary/update.py [hunk 1]
```diff
@@ -764,6 +771,345 @@ def confirm_import_gloss_csv(request):
...
+
+            return render(request, 'dictionary/import_nzsl_share_gloss_csv_confirmation.html',
+                          {'glosses_new': new_glosses,
+                           'dataset': dataset, })
+        else:
+            # If form is not valid, set a error message and return to the original form.
+            messages.add_message(request, messages.ERROR, _('The provided CSV-file does not meet the requirements '
+                                                            'or there is some other problem.'))
+            return render(request, 'dictionary/import_nzsl_share_gloss_csv.html', {'import_csv_form': form}, )
+    else:
```
Comment (jadedarko) [inline signbank/dictionary/update.py]: **suggestion**: because the GET usecase is so much shorter, you may want to consider putting it first, then returning, and un-indenting the POST usecase (i.e.: no `else:`)

The same applies to the returns for invalid form / exceptions

This article is something that I keep coming back to: [align the happy path to the left edge](https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88)

File: signbank/dictionary/update.py [hunk 2]
```diff
@@ -764,6 +771,345 @@ def confirm_import_gloss_csv(request):
...
+                # Flush request.session['glosses_new'] and request.session['dataset']
+                del request.session['glosses_new']
+                del request.session['dataset_id']
+
+                # Set a message to be shown so that the user knows what is going on.
+                # messages.add_message(request, messages.SUCCESS, _('Glosses were added succesfully.'))
+            return render(request, "dictionary/import_nzsl_share_gloss_csv_confirmation.html", {'glosses_added': glosses_added,
+                                                                                     'dataset': dataset.name})
+        else:
+            return HttpResponseRedirect(reverse('dictionary:import_nzsl_share_gloss_csv'))
```
Comment (jadedarko) [inline signbank/dictionary/update.py]: **suggestion**: as above, you may want to consider the error case returns first, so that the main functionality doesn't need as much indentation

File: signbank/dictionary/update.py [hunk 3]
```diff
@@ -764,6 +771,345 @@ def confirm_import_gloss_csv(request):
...
+                    delimiter=',',
+                    quotechar='"'
+                )
+                for row in glossreader:
+                    if glossreader.line_num == 1:
+                        continue
+                    new_glosses.append(row)
+            except csv.Error as e:
+                # Can't open file, remove session variables
+                if 'dataset_id' in request.session: del request.session['dataset_id']
```
Comment (jadedarko) [inline signbank/dictionary/update.py]: **polish**: `request.session.pop('dataset_id', None)` may be clearer

File: signbank/dictionary/update.py [hunk 4]
```diff
@@ -764,6 +771,345 @@ def confirm_import_gloss_csv(request):
...
+                                comment=comment_content,
+                                site=site,
+                                is_public=False,
+                                submit_date=comment_submit_date
+                            ))
+
+                    if gloss_data.get("videos", None):
+                        for i, video_url in enumerate(gloss_data["videos"].split("|")):
+                            extension = video_url[-3:]
+                            file, _ = urlretrieve(
```
Comment (jadedarko) [inline signbank/dictionary/update.py]: **issue**: retrieving all the assets will likely take a significant time. If possible, this should be moved to a background task instead of occurring in the request.

**Review by jadedarko — CHANGES_REQUESTED** [review]
I've added a few comments, but the main thing I am concerned with is retrieving the assets during the request.

Is it possible to move this to a background task? It's likely that retrieving the actual asset files is the reason the request is timing out.

The project does not currently have celery, but it would be good if we could add it even just for this usecase.

File: signbank/dictionary/tasks.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        # Save the file into the new path.
+        saved_file_path = glossvideo.videofile.storage.save(full_new_path, old_file)
+        # Set the actual file path to videofile.
+        glossvideo.videofile = saved_file_path
+    return glossvideo
+
+
+# background task schedule kicks task off 30 seconds after call
+@background(schedule=30)
+def retrieve_videos_for_glosses(video_details, gloss_pks):
```
Comment (jadedarko) [inline signbank/dictionary/tasks.py]: **suggestion**: I know the existing style doesn't use type annotation, but it would be helpful to have that here - the one usage has these arguments in the opposite order

Comment (jadedarko) [reply signbank/dictionary/tasks.py]: Further to this, because the pks for the glosses are in the list of video dicts, it may be better to extract those from the dictionaries. Otherwise there is the potential that the list of pks might not correspond to the pks in the list of dicts.

File: signbank/dictionary/tasks.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
+from urllib.request import urlretrieve
+
+from background_task import background
+from django.conf import settings
+
+from .models import FieldChoice, Gloss
+from ..video.models import GlossVideo
+
+
+def move_glossvideo_to_valid_filepath(glossvideo):
```
Comment (jadedarko) [inline signbank/dictionary/tasks.py:19]: **praise**: nicely done on catching this

File: signbank/dictionary/update.py [hunk 7]
```diff
@@ -764,6 +774,316 @@ def confirm_import_gloss_csv(request):
...
+                        idgloss=f"{gloss['word']}_row{row_num}",
+                        idgloss_mi=gloss.get("maori", None),
+                        notes=gloss.get("notes", ""),
+                        created_by=request.user,
+                        updated_by=request.user
+                    ))
+
+                bulk_created = Gloss.objects.bulk_create(bulk_create_gloss)
+                idglosses = [x.idgloss for x in bulk_created]
+                bulk_created_with_pk = Gloss.objects.filter(
```
Comment (jadedarko) [inline signbank/dictionary/update.py]: **suggestion**: we shouldn't need to retrieve these again - every currently supported version of postgres [should retrieve the pk on bulk_create](https://docs.djangoproject.com/en/5.0/ref/models/querysets/#bulk-create)

fwiw we're running 14 in prod & the docker compose

File: signbank/dictionary/update.py [hunk 8]
```diff
@@ -764,6 +774,316 @@ def confirm_import_gloss_csv(request):
...
+                    idgloss__in=idglosses
+                )
+
+                gloss_pks = []
+
+                for gloss in bulk_created_with_pk:
```
Comment (jadedarko) [inline signbank/dictionary/update.py:925]: **suggestion**: this method is quite long by this point, might be worth splitting it up

File: signbank/dictionary/update.py [hunk 9]
```diff
@@ -764,6 +774,316 @@ def confirm_import_gloss_csv(request):
...
+                    # will iterate over these glosses again after bulk creating and to make sure
+                    # we preserve order
+                    new_glosses[str(row_num)] = gloss
+                    bulk_create_gloss.append(Gloss(
+                        dataset=dataset,
+                        # need to make idgloss unique in dataset,
+                        # but gloss word can appear in multiple rows, so
+                        # idgloss will be updated to word:pk in second step
+                        idgloss=f"{gloss['word']}_row{row_num}",
+                        idgloss_mi=gloss.get("maori", None),
```
Comment (jadedarko) [inline signbank/dictionary/update.py]: **question**: is it not possible to also have the maori gloss word appear in multiple rows?

File: signbank/dictionary/update.py [hunk 10]
```diff
@@ -764,6 +774,316 @@ def confirm_import_gloss_csv(request):
...
+                site = Site.objects.get_current()
+                comment_submit_date = datetime.datetime.now(tz=get_current_timezone())
+                semantic_fields = FieldChoice.objects.filter(
+                    field="semantic_field"
+                ).values_list("english_name", "pk")
+                semantic_fields_dict = {field[0]: field[1] for field in semantic_fields}
+
+                for row_num, gloss in enumerate(request.session['glosses_new']):
+                    # will iterate over these glosses again after bulk creating and to make sure
+                    # we preserve order
```
Comment (jadedarko) [inline signbank/dictionary/update.py]: **question**: if the row numbers are being used to preserve order, should the iteration over glosses not iterate over the `new_glosses` dict, and then retrieve the corresponding instance from the bulk create results?

File: signbank/dictionary/update.py [hunk 11]
```diff
@@ -764,6 +774,316 @@ def confirm_import_gloss_csv(request):
...
+                idglosses = [x.idgloss for x in bulk_created]
+                bulk_created_with_pk = Gloss.objects.filter(
+                    idgloss__in=idglosses
+                )
+
+                gloss_pks = []
+
+                for gloss in bulk_created_with_pk:
+                    gloss_pks.append(gloss.pk)
+                    old_id_gloss = gloss.idgloss.split("_row")
```
Comment (jadedarko) [inline signbank/dictionary/update.py]: **suggestion**: since it's assumed to be a 2-tuple, you could label the parts by assigning them to variables here instead of using the indices:

```
word_en, row_number = ...
```

File: signbank/dictionary/update.py [hunk 12]
```diff
@@ -764,6 +774,316 @@ def confirm_import_gloss_csv(request):
...
+
+                # Bulk create entities related to the gloss, and bulk update the glosses' idgloss
+                Comment.objects.bulk_create(comments)
+                GlossTranslations.objects.bulk_create(translations)
+                Gloss.objects.bulk_update(bulk_update_glosses, ["idgloss", "idgloss_mi"])
+                Gloss.semantic_field.through.objects.bulk_create(bulk_semantic_fields)
+
+                # call background task to retrieve videos for glosses and activate the task
+                # processor for 10 minutes
+                retrieve_videos_for_glosses(gloss_pks, videos)
```
Comment (jadedarko) [inline signbank/dictionary/update.py]: **issue**: the parameters for this appear to be swapped

**Review by jadedarko — COMMENTED** [review]
This is looking great, thank you for coming up with a stopgap solution for background processing 🙏🏼

File: signbank/dictionary/update.py (hunk 9 above)
Comment (nora-ns) [reply signbank/dictionary/update.py]: In theory yes. However there is a unique constraint on (dataset, idgloss) which is why I only added it to the idgloss

File: signbank/dictionary/update.py (hunk 10 above)
Comment (nora-ns) [reply signbank/dictionary/update.py]: I updated the comment, I think it was only my initial thought that I needed to preserve order. 
What it actually is is creating an association between the Gloss we're about to create and the gloss_data via the row number

File: signbank/dictionary/update.py (hunk 9 above)
Comment (nora-ns) [reply signbank/dictionary/update.py]: I also looked at the production export csv from NZSL Share again and 'maori' might actually be a comma separated list.

File: signbank/dictionary/update.py (hunk 7 above)
Comment (nora-ns) [reply signbank/dictionary/update.py]: I always get this mixed up

**Comment by nora-ns** [comment]
@jadedarko I cleaned this up and added tests.
I also replaced `django-background-tasks` with just threads because I must not have tested in correctly and calling the processor in the request cause a time-out again.
Downside is that the tasks now does not seem to retrieve Gloss querysets anymore, but single glosses can be retrieved 🤷🏻‍♀️

File: signbank/settings/base.py [hunk 13]
```diff
@@ -243,6 +243,8 @@
 else:
     GLOSS_VIDEO_FILE_STORAGE = 'signbank.video.models.GlossVideoStorage'
 
+NZSL_SHARE_HOSTNAME = os.getenv('NZSL_SHARE_HOSTNAME')
```
Comment (nora-ns) [inline signbank/settings/base.py:246]: Reminder for myself to look up where this new environment variable actually needs to be added for UAT and Prod

File: signbank/dictionary/tests/tests_tasks.py [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+        }]
+
+        with mock.patch("signbank.dictionary.tasks.urlretrieve") as mock_retrieve:
+            mock_retrieve.return_value = (dummy_file, None)
+            retrieve_videos_for_glosses(video_details)
+            mock_retrieve.assert_called_once()
+
+        videos = GlossVideo.objects.filter(gloss=self.gloss)
+        self.assertTrue(videos.exists())
+        self.assertEqual(videos.count(), 1)
```
Comment (jadedarko) [inline signbank/dictionary/tests/tests_tasks.py:52]: **question**: do we want to test the filename etc. for this? `video_details` could be used to assert the task correctly assigns those values

**Review by jadedarko — APPROVED** [review]
Thank you for this!

Great work on using threading instead of a background processor; that's such a simple solution for when scheduling isn't required

**Review by jadedarko — CHANGES_REQUESTED** [review]
Reading over [this](https://docs.djangoproject.com/en/4.2/ref/databases/#caveats) (specifically, the last line in that section), it seems  you'll need to handle closing the database connection at the end of your function

I was alerted to that from this [stack overflow answer](https://stackoverflow.com/a/28913218), and it's backed up by [this issue](https://code.djangoproject.com/ticket/9878#comment:4) which was closed as wontfix

[Here's an example](https://github.com/slackapi/bolt-python/pull/281/files#r606142527) of how it's been addressed in other projects

**Comment by nora-ns** [comment]
Good find. I've added in the `connections.close()` at the end of the task

**Review by jadedarko — APPROVED** [review]

File: signbank/dictionary/update.py [hunk 15]
```diff
@@ -764,6 +776,386 @@ def confirm_import_gloss_csv(request):
...
+                            object_pk=gloss.pk,
+                            user_name=user_name,
+                            comment=comment_content,
+                            site=site,
+                            is_public=False,
+                            submit_date=comment_submit_date
+                        ))
+
+                # Add ShareValidationAggregation
+                bulk_share_validation_aggregations.append(ShareValidationAggregation(
```
Comment (jadedarko) [inline signbank/dictionary/update.py:1057]: **question**: are the agree/disagree counts ever likely to want updating?

**Comment by DSRU** [comment]
Probably not. The agree/disagree counts at the time of import into Signbank gives us an idea of which signs to take further for the validation process. If there are changes to this (or further comments) on NZSL Share afterwards, then those would not be easy to incorporate, because we ignore NZSL Share signs that have already been imported.

A more complete integration of NZSL Share and Signbank is out of scope for now.

**Review by jadedarko — APPROVED** [review]
Thank you for your thoroughness in implementing this!

Spent more time than expected reading back over it, but I believe this is good to go!

Regarding the comments that Simeon had during the guild meeting, it _might_ be worth reading over [this resource](https://superfastpython.com/threading-in-python/) if you have the time

My take is that this use-case is well suited to threads, so your approach is good!
